src/Perpetuum/Services/EventServices/EventProcessors/OreNPCSpawner.cs

```diff
@@ -0,0 +1,199 @@
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Perpetuum.Services.EventServices.EventProcessors
+{
+    /// <summary>
+    /// EventListener for each zone, receives messages for mineralnode mined and spawns npc presence based on the INpcReinforcementsRepository configurations
+    /// </summary>
+    public class OreNpcSpawner : EventProcessor<EventMessage>
+    {
+        private const int SPAWN_DIST_FROM_FIELD = 50;
+        private readonly TimeSpan ORE_SPAWN_LIFETIME = TimeSpan.FromHours(3);
+        private readonly TimeSpan SPAWN_DELAY = TimeSpan.FromSeconds(10);
+
+        private readonly IZone _zone;
+        private readonly IDictionary<MineralNode, DynamicPresence> spawnedPresences = new Dictionary<MineralNode, DynamicPresence>();
+        private readonly IDictionary<MineralNode, INpcReinforcements> reinforcementsByMineralNode = new Dictionary<MineralNode, INpcReinforcements>();
+        private readonly INpcReinforcementsRepository _npcReinforcementsRepo;
+        private readonly IEnumerable<IMineralConfiguration> _mineralConfigs;
+
+        public OreNpcSpawner(IZone zone, INpcReinforcementsRepository reinforcementsRepo, IMineralConfigurationReader mineralConfigurationReader)
+        {
+            _zone = zone;
+            _npcReinforcementsRepo = reinforcementsRepo;
+            _mineralConfigs = mineralConfigurationReader.ReadAll().Where(c => c.ZoneId == zone.Id);
+        }
+
+        private void OnPresenceExpired(Presence presence)
+        {
+            var matchedEntries = spawnedPresences.Where(p => p.Value.Equals(presence)).ToList();
+            foreach (var pair in matchedEntries)
+            {
+                pair.Value.PresenceExpired -= OnPresenceExpired;
+                spawnedPresences.Remove(pair.Key);
+            }
+        }
+
+        private void RemoveEntry(MineralNode node)
+        {
+            if (spawnedPresences.ContainsKey(node))
+            {
+                if (spawnedPresences[node] != null)
+                {
+                    spawnedPresences[node].PresenceExpired -= OnPresenceExpired;
+                }
+                spawnedPresences.Remove(node);
+            }
+        }
+
+        private Position TryFindSpawnLocation(Position start, double range)
+        {
+            var p = new Position();
+            List<Point> result = null;
+            for (int i = 0; i < 10; i++)
+            {
+                var random = FastRandom.NextDouble(0.0, 1.0);
+                var pos = start.OffsetInDirection(random, range);
+                var posFinder = new ClosestWalkablePositionFinder(_zone, pos);
+                posFinder.Find(out p);
+                result = _zone.FindWalkableArea(p, _zone.Size.ToArea(), 100);
+                if (result != null)
+                {
+                    break;
+                }
+            }
+            if (result == null)
+            {
+                return Position.Empty;
+            }
+            return p;
```

Comment:
```
    if (result != null) return p;
  }
  return Position.Empty;
}
```

Reply:
Is there an optimization here? Or stylistic/convention?
IE: check the desired-state case first and return, let fall-through return the empty case?

Reply:
Oh! I see I can just return from that loop.. duh!  Sorry I misread your suggestion.